### PR TITLE
Implementation of Most Similar Neighbor (MSN)

### PR DESCRIPTION
**Pull request by grovduck**
This PR implements Most Similar Neighbor (MSN) (Moeur and Stage, 1998) which uses Canonical Correlation Analysis to associate `X` and `y` matrices.  Linear coefficients for both the `X` and `y` matrices are calculated such that the calculated fitted scores of each matrix maximize the correlation between matrices.

This implementation generally follows the code presented for method `msn` in `yaImpute`.  I was unable to directly port the R function `cancor` into Python.  The R implementation uses both QR decomposition and SVD decomposition of matrices to arrive at the `X` and `y` coefficients and the QR decomposition can introduce reordering of columns that I was unable to reproduce in Python.  I finally stumbled on `statsmodels.CanCorr` which uses SVD decomposition (exclusively) but throws an error if the matrices do not have full rank after the decomposition.  I have borrowed heavily from this implementation and we need to make sure we are adhering to their license.  Unfortunately, I wasn't able to figure out how to use sklearn's own [`CCA`](https://scikit-learn.org/stable/modules/generated/sklearn.cross_decomposition.CCA.html) to produce the correct output.  This may be a place for further follow-up.

Likewise, there is a test (`ftest_cor`) ported directly from the `yai` function in `yaImpute` which (presumably) finds the significant CCorA axes to use based on p-values from an F-test (I think I have this right?).  This requires us to include `scipy` as a requirement, which is already a requirement of `scikit-learn`.

* Moeur, M. and Stage, A.R., 1995. Most similar neighbor: an improved sampling inference procedure for natural resource planning. Forest science, 41(2), pp.337-359.

### CODE REVIEW

**Review by aazuspan — COMMENTED** [review]
Sorry, this review is a bit of a bummer! All the code looks great, but I think we have some work to do to be compliant with the licenses of `statsmodels` and (especially) `yaImpute`. 

I'm not sure we need to get everything sorted out right now, and maybe we leave this for a final step before the first release, since some of this code will potentially change before then. 

Let me know what you think. If you're okay with it, I'd be fine to hold off on getting into the nitty gritty of licensing and just make sure we're noting in every docstring where code is ported or adapted from, so that we can come back to it later.

**Comment by grovduck** [comment]
> Sorry, this review is a bit of a bummer! All the code looks great, but I think we have some work to do to be compliant with the licenses of statsmodels and (especially) yaImpute.

No bummer at all - this is really important to address and I agree that we need to figure this out.  I just took a bit of a deep dive myself into `yaImpute` licensing history and it looks like they are GPL v3 because they relied on libraries that were GPL v3 (specifically ANN) - see [here](https://cran.r-project.org/web/packages/yaImpute/COPYRIGHTS).

You might be able to answer this question, though.  When we use another package and call its functionality directly, are we beholden to their licensing in the same way that would be required by porting?  For example, `scikit-learn` uses BSD-3 licensing, but because we're *using* this code rather than *porting* this code, are the terms different?

> I'm not sure we need to get everything sorted out right now, and maybe we leave this for a final step before the first release, since some of this code will potentially change before then. Let me know what you think. If you're okay with it, I'd be fine to hold off on getting into the nitty gritty of licensing and just make sure we're noting in every docstring where code is ported or adapted from, so that we can come back to it later.

I agree that we probably don't want to high-center on this right now, but we definitely need to create an issue (possibly associated with a milestone!) that tracks this.  That's a really good idea to put markers in docstrings where we've ported code from others - that would be a first step before including the actual license information.  

This is a new world to me and it all seems a bit squishy.  But erring on the side of caution definitely seems appropriate.

File: src/sknnr/transformers/_ccora.py
Comment (grovduck) [reply src/sknnr/transformers/_ccora.py:6]: I agree with everything you've noted here, including the steps to be compliant.  If we did want to avoid having to use the GPL v3 license, I'm not totally clear how we would re-implement everything from scratch.  There probably aren't too many ways to do the same thing (at least not well) and one might still use the original work as inspiration.

I don't know how much you've been using (or trust!) ChatGPT, but I found the following Q & A pretty interesting:

> Q. when you use functions from another python package in your code, do you have to abide by their licensing restrictions?  For example, if you use a library that is licensed as GPL v3, can your project be BSD v3?
> 
> A. When you use functions or libraries from another Python package in your code, the licensing restrictions of the library you are using generally apply to your project as well. This means that if you use a library licensed under the GPL v3, your project needs to comply with the terms of the GPL v3 license.
> 
> The GPL is a copyleft license that requires any derivative works or projects that use GPL-licensed code to also be licensed under the GPL or a compatible license. This "viral" nature of the GPL ensures that the freedoms and obligations of the GPL license are preserved in derivative works.
> 
> If you want to license your project under a different license, such as the BSD v3 license, and you are using a GPL-licensed library, there could be a conflict. The GPL v3 is generally not compatible with the BSD v3 license because it imposes more stringent requirements on derivative works.
> 
> To address this, you have a few options:
> 
> 1. Use a Compatible Library: Look for alternative libraries that are licensed under a more permissive license, such as the BSD v3 license. This way, you can avoid the licensing conflict altogether.
> 
> 2. Dual Licensing: Contact the library's author or copyright holder and inquire about the possibility of obtaining a separate license for your use case. They may grant you a special license or provide an alternative license option that is compatible with your desired licensing terms.
> 
> 3. Separate Execution: If your project can be structured in a way that the GPL-licensed library is a separate component and is not directly linked to or integrated into your code, you may be able to avoid the requirement of applying the GPL to your entire project. In this case, the GPL-licensed code would be considered a separate program, and the licensing obligations would apply only when distributing or modifying the GPL-licensed component.
> 
> It's important to consult the specific terms and conditions of the library's license to understand your obligations. Licensing issues can be complex, so seeking legal advice or discussing with the library's maintainers or a licensing expert can provide further clarity in navigating these situations.
> 
So, you've mentioned the possibility of 1 which makes sense, although I believe the world of kNN software is pretty narrow!  I could probably get Andy or Nick (`yaImpute` developers) to allow us to have a dual license (2), but they are probably bound by the ANN library and couldn't release it in that way.  I think 3 is a pretty interesting approach, although much of what we're doing is more or less a port of `yaImpute`, so I don't know if that would be worth the effort.

Sorry, just more questions than answers!

File: src/sknnr/transformers/_ccora.py
Comment (grovduck) [reply src/sknnr/transformers/_ccora.py:42]: This sounds fine to me.  Many fewer decisions than the GPL case.  And I agree on the additional documentation in a docstring to relay why we've done it this way.  I really prefer that we could get sklearn's `CCA` to work, but that's for another day.

What do you think about adding the additional detail to the docstring as part of this PR, but punt the licensing until we deal with it all as a whole (as you say, before the first release)?

**Comment by aazuspan** [comment]
> I just took a bit of a deep dive myself into yaImpute licensing history and it looks like they are GPL v3 because they relied on libraries that were GPL v3 (specifically ANN) - see [here](https://cran.r-project.org/web/packages/yaImpute/COPYRIGHTS).

Good point! It looks like this will depend on which files we ported from then, since they seem to give a very permissive license for anything outside of the ANN code:

```
=====All other files including src/rfoneprox.cpp ==
All files in this package, other than those refered to above, were prepared by 
Nicholas L. Crookston, a U.S. Government employee working on official time 
and are in the public domain. 
```

So it may be worth noting in docstrings which files were ported, since that will determine the appropriate license.

> When we use another package and call its functionality directly, are we beholden to their licensing in the same way that would be required by porting? For example, scikit-learn uses BSD-3 licensing, but because we're using this code rather than porting this code, are the terms different?

No, at least in the case of BSD-3 and MIT there are no restrictions on usage of code, only on redistribution or modification. I think GPL might be a little trickier in this area, and it seems like it's open to interpretation around whether GPL allows "dynamic linking" and whether importing in Python even counts as "dynamic linking".

> I agree that we probably don't want to high-center on this right now, but we definitely need to create an issue (possibly associated with a milestone!) that tracks this. That's a really good idea to put markers in docstrings where we've ported code from others - that would be a first step before including the actual license information.

Sounds good! For this PR, can you add those markers for the code included here? After that, I'm good to merge!

> This is a new world to me and it all seems a bit squishy. But erring on the side of caution definitely seems appropriate.

Yes, squishy is definitely my impression too! I think the truth is that very few projects follow all the rules correctly, and there's virtually zero chance that anyone notices or cares. We're probably going above and beyond with this level of effort.

**Comment by grovduck** [comment]
> Sounds good! For this PR, can you add those markers for the code included here? After that, I'm good to merge!

@aazuspan, can you check on the markers I left?  Two issues:
1. I used `REFERENCE` as a marker to be able to refer to these at a later point
2. I used permalinks for derived lines of codes which exceed line length so I left `#noqa: E501`s at the ends of these lines

Neither of these decisions am I wedded to, so please suggest alternatives if you'd like.

File: src/sknnr/transformers/_ccora.py
Comment (aazuspan) [reply src/sknnr/transformers/_ccora.py:44]: Getting `CCA` to work would be great, but I agree that can wait. 

I'm happy with this plan!

**Review by aazuspan — APPROVED** [review]
> I used REFERENCE as a marker to be able to refer to these at a later point

We could use a [notes and references](https://numpydoc.readthedocs.io/en/latest/format.html#notes) section following numpydoc, but if these are just placeholders then I'm fine with it as is.

> I used permalinks for derived lines of codes which exceed line length so I left #noqa: E501s at the ends of these lines

I never like having to do this, but I'm not sure there's any better alternative. Sphinx does have a way to [cross-reference documentation from other projects](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html) and I imagine there's something similar for mkdocs, but I think a permalink is probably a better option in case their implementation or license changes.

I'm good to merge if you are!

File: src/sknnr/transformers/_ccora.py
Comment (aazuspan) [reply src/sknnr/transformers/_ccora.py]: > I don't know how much you've been using (or trust!) ChatGPT

I was definitely consulting with Google Bard trying to figure out how to comply with these licenses!

>  I could probably get Andy or Nick (yaImpute developers) to allow us to have a dual license (2), but they are probably bound by the ANN library and couldn't release it in that way.

Another interesting facet here is that the [ANN library](http://www.cs.umd.edu/~mount/ANN/) is distributed under [GNU **Lesser** Public License v2](http://www.cs.umd.edu/~mount/ANN/Files/1.1.2/License_1.1.2.txt), which I believe allows you to distribute either under LGPL or a more restrictive GPL. So maybe we could get a dual license to avoid GPL v3, but would still be bound by LGPL, which probably isn't worth the effort...

Comment (grovduck) [reply src/sknnr/transformers/_ccora.py]: I'm going to email Nick Crookston and Andy Finley on this to see what they might recommend.  My impression is that the GPL license on yaImpute came after they were actively involved.  Whether that was because the new maintainer sought to make it GPL v3 or if they were trying to be compliant, I'm not sure.  

That's pretty interesting about ANN.  I think we both concur that we'd like to be a bit less restrictive than GPL v3.  Any word I get back from Nick or Andy, I'll post in #29.

**Comment by grovduck** [comment]
> ... if these are just placeholders then I'm fine with it as is.

Yes, I think these are just placeholders and will eventually be removed.  I'm going to merge this one to keep things going, but I've left a note in [#29](https://github.com/lemma-osu/scikit-learn-knn-regression/issues/29#issuecomment-1569130049) to remind us to deal with these.

**Pull request merged**
